flow/display_list_utils.h

```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
+    FML_DISALLOW_COPY_AND_ASSIGN(SaveLayerData);
   };
 
-  std::vector<std::unique_ptr<SaveInfo>> saved_infos_;
+  std::vector<std::unique_ptr<LayerData>> layer_infos_;
+
+  // A drawing operation that is not geometric in nature (but which
+  // may still apply a MaskFilter - see |kApplyMaskFilter| below).
+  static constexpr int kIsNonGeometric = 0x00;
+
+  // A geometric operation that is defined as a fill operation
+  // regardless of what the current paint Style is set to.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsFilledGeometry = 0x01;
+
+  // A geometric operation that is defined as a stroke operation
+  // regardless of what the current paint Style is set to.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsStrokedGeometry = 0x02;
+
+  // A geometric operation that may be a stroke or fill operation
+  // depending on the current state of the paint Style attribute.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsDrawnGeometry = 0x04;
+
+  static constexpr int kIsAnyGeometryMask =  //
+      kIsFilledGeometry |                    //
+      kIsStrokedGeometry |                   //
+      kIsDrawnGeometry;
+
+  // A geometric operation which has a path that might have
+  // end caps that are not rectilinear which means that square
+  // end caps might project further than half the stroke width
+  // from the geometry bounds.
+  // A rectilinear path such as |drawRect| will not have
+  // diagonal end caps. |drawLine| might have diagonal end
+  // caps depending on the angle of the line, and more likely
+  // |drawPath| will often have such end caps.
+  static constexpr int kGeometryMayHaveDiagonalEndCaps = 0x08;
+
+  // A geometric operation which has joined vertices that are
+  // not guaranteed to be smooth (angles of incoming and outgoing)
+  // segments at some joins may not have the same angle) or
+  // rectilinear (squares have right angles at the corners, but
+  // those corners will never extend past the bounding box of
+  // the geometry pre-transform).
+  // |drawRect|, |drawOval| and |drawRRect| all have well
+  // behaved joins, but |drawPath| might have joins that cause
+  // mitered extensions outside the pre-transformed bounding box.
+  static constexpr int kGeometryMayHaveProblematicJoins = 0x10;
+
+  // Some operations are inherently non-geometric and yet have the
+  // mask filter applied anyway.
+  // |drawImage| variants behave this way.
+  static constexpr int kApplyMaskFilter = 0x20;
+
+  // In very rare circumstances the ImageFilter is ignored.
+  // This is one of the few flags that turns off a step in
+  // estimating the bounds, rather than turning on any steps.
+  static constexpr int kIsUnfiltered = 0x40;
+
+  static constexpr SkScalar kMinStrokeWidth = 0.01;
+
+  skstd::optional<SkBlendMode> blend_mode_ = SkBlendMode::kSrcOver;
+  sk_sp<SkColorFilter> color_filter_;
```

Comment:
Is this safe to collect on non-raster task runners, and if not are we certain it will always get collected on the raster task runner?

Reply:
Ditto to above. I can't find where we've ever held these other than in an sk_sp...?

Reply:
I suspect the `SkColorFilter` is safe to release on a non-GPU thread, since I don't think it ever ends up holding any GPU related resources.

Reply:
`SkColorFilters` are `SkFlattenable`'s. If they held GPU resources, they could be flattened. So I think this is safe. They are also [documented to be shareable across threads](https://github.com/google/skia/blob/941812fdcc8426b29aca8072326cbeb9371cc9a3/include/core/SkColorFilter.h#L25).

Reply:
I think so too. And if we get rid of the vestigial code that attempts to create an `SkImageImageFilter` then the same could be said for our `SkImageFilter` uses.